src/lightning/fabric/strategies/xla_fsdp.py

```diff
@@ -0,0 +1,404 @@
+        """
+        if not _TORCH_GREATER_EQUAL_2_0:
+            raise NotImplementedError(
+                "Saving and loading checkpoints with the `XLAFSDPStrategy` is not supported in PyTorch < 2.0."
+                " Please upgrade `torch` or file an issue: `https://github.com/Lightning-AI/lightning/issues`."
+            )
+        if storage_options is not None:
+            raise TypeError(
+                "`XLAFSDPStrategy.save_checkpoint(..., storage_options=...)` is not supported because"
+                " `XLAFSDPStrategy` does not use the `CheckpointIO`."
+            )
+        # broadcast the path from rank 0 to ensure all the states are saved in a common path
+        path = self.broadcast(path)
+        if Path(path).is_dir() and os.listdir(path):
+            raise FileExistsError(f"The checkpoint directory already exists and is not empty: {path}")
+        from torch_xla.distributed.fsdp import XlaFullyShardedDataParallel as XLAFSDP
+
+        modules = [module for module in state.values() if isinstance(module, XLAFSDP)]
+        if len(modules) == 0:
+            raise ValueError(
+                "Could not find a XLA FSDP model in the provided checkpoint state. Please provide the model as"
+                " part of the state like so: `save_checkpoint(..., state={'model': model, ...})`. Make sure"
+                " you set up the model (and optimizers if any) through the strategy before saving the checkpoint."
+            )
+        if len(modules) > 1:
+            raise ValueError(
+                "Found multiple XLA FSDP modules in the given state. Saving checkpoints with FSDP is"
+                " currently limited to a single model per checkpoint. To save multiple models, call the"
+                " save method for each model separately with a different path."
+            )
+        import torch_xla.core.xla_model as xm
+
+        rank = xm.get_ordinal()
+        world_size = xm.xrt_world_size()
+
+        converted_state: Dict[str, Any] = {}
+        for key, obj in state.items():
+            # convert the state
+            if isinstance(obj, Module) and isinstance(obj, XLAFSDP):
+                converted = obj.state_dict()
+                # add shard_metadata to state
+                converted_state["shard_metadata"] = obj.get_shard_metadata()
+            elif isinstance(obj, Optimizer):
+                converted = obj.state_dict()
+            else:
+                converted = obj
+            _apply_filter(key, filter or {}, converted, converted_state)
+
+        self.checkpoint_io.save_checkpoint(
+            converted_state, f"{path}_rank-{rank:08d}-of-{world_size:08d}.pth", storage_options=storage_options
+        )
+
+        if self._state_dict_type == "full":
+            from torch_xla.distributed.fsdp import consolidate_sharded_model_checkpoints
+
+            if xm.is_master_ordinal(local=False):
+                consolidate_sharded_model_checkpoints(
```

Comment:
Shouldn't this path remove the sharded checkpoints after consolidation? Then in `remove_checkpoint` you would only remove the sharded pieces in the `state_dict_type != full` case

Reply:
This makes sense. Will add